### PR TITLE
feat: allow folder uploads

### PR DESCRIPTION
**Pull request by mxschmitt**
Motivation: This adds support for `input[webkitdirectory]` folder uploads. For local mode, we just pass the directory to the browser, for "remote mode" we stream the files to the server and keeping the same parent directory name. This one is also used inside the browser in the `webkitRelativePath` field in the `File` type.

In Chromium headless, there is [a bug](https://issues.chromium.org/issues/345393164), that folder uploads won't consider nested files / is this a blocker for the feature to land?
~~Firefox + WebKit implementation have been landed and look like they are working. Requires rolls, hence the tests are still not passing.~~

Feature request: https://github.com/microsoft/playwright/issues/6854

### CODE REVIEW

File: docs/src/api/class-locator.md [hunk 1]
```diff
@@ -2164,6 +2164,7 @@ When all steps combined have not finished during the specified [`option: timeout
 * since: v1.14
 
 Upload file or multiple files into `<input type=file>`.
+In order to upload a directory (`[webkitdirectory]`) pass the path to the directory.
```
Comment (dgozman) [inline docs/src/api/class-locator.md:2167]:
```suggestion
For inputs with a `[webkitdirectory]` attribute, only a single directory path is supported.
```

File: packages/playwright-core/src/server/dom.ts [hunk 2]
```diff
@@ -633,7 +633,7 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
         return;
       if (element.tagName !== 'INPUT')
         throw injected.createStacklessError('Node is not an HTMLInputElement');
-      if (multiple && !(element as HTMLInputElement).multiple)
+      if (multiple && !(element as HTMLInputElement).multiple && !(element as HTMLInputElement).webkitdirectory)
```
Comment (dgozman) [inline packages/playwright-core/src/server/dom.ts:636]: Let's throw a similar error when you pass multiple files into directory inputs.

File: packages/playwright-core/src/server/dom.ts [hunk 3]
```diff
@@ -647,7 +647,12 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
         await Promise.all(localPaths.map(localPath => (
           fs.promises.access(localPath, fs.constants.F_OK)
         )));
+        const isDirectoryUpload = localPaths.length === 1 ? (await fs.promises.stat(localPaths[0])).isDirectory() : false;
+        const waitForChangeEvent = isDirectoryUpload ? this.evaluateInUtility(([_, node]) => new Promise<any>(fulfil => {
```
Comment (dgozman) [inline packages/playwright-core/src/server/dom.ts:651]: typo: `fulfil` -> `fulfill`

File: packages/playwright-core/src/server/dom.ts [hunk 4]
```diff
@@ -647,7 +647,12 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
         await Promise.all(localPaths.map(localPath => (
           fs.promises.access(localPath, fs.constants.F_OK)
         )));
+        const isDirectoryUpload = localPaths.length === 1 ? (await fs.promises.stat(localPaths[0])).isDirectory() : false;
```
Comment (dgozman) [inline packages/playwright-core/src/server/dom.ts:650]: Perhaps only check this when `webkitdirectory` attribute is present?

File: packages/playwright-core/src/client/elementHandle.ts [hunk 5]
```diff
@@ -262,16 +262,42 @@ export async function convertInputFiles(files: string | FilePayload | string[] |
   if (items.some(item => typeof item === 'string')) {
     if (!items.every(item => typeof item === 'string'))
       throw new Error('File paths cannot be mixed with buffers');
+    const directoryCount = (await Promise.all(items.map(async item => (await fs.promises.stat(item as string)).isDirectory()))).filter(Boolean).length;
+    if (directoryCount > 1)
+      throw new Error('Only one directory can be uploaded at a time');
```
Comment (dgozman) [inline packages/playwright-core/src/client/elementHandle.ts:267]: Let's be more strict - either a single directory, or a list of files, but not files+directory.

File: packages/playwright-core/src/client/elementHandle.ts [hunk 6]
```diff
@@ -262,16 +262,42 @@ export async function convertInputFiles(files: string | FilePayload | string[] |
...
-        const lastModifiedMs = (await fs.promises.stat(item)).mtimeMs;
-        const { writableStream: stream } = await context._wrapApiCall(() => context._channel.createTempFile({ name: path.basename(item), lastModifiedMs }), true);
-        const writable = WritableStream.from(stream);
-        await pipelineAsync(fs.createReadStream(item), writable.stream());
-        return stream;
-      }));
-      return { streams };
+      let streams: channels.WritableStreamChannel[] | undefined;
+      let localPaths: string[] | undefined;
+      for (const item of items as string[]) {
```
Comment (dgozman) [inline packages/playwright-core/src/client/elementHandle.ts:272]: I think we should keep `Promise.all()` and extract a method here. It is getting complicated.

File: packages/protocol/src/protocol.yml [hunk 7]
```diff
@@ -1191,6 +1191,22 @@ BrowserContext:
       returns:
         writableStream: WritableStream
 
+    createTempDirectory:
```
Comment (dgozman) [inline packages/protocol/src/protocol.yml:1194]: Let's merge this and previous methods and call it `createTempFiles`, giving an optional `rootDirName` parameter?

File: packages/protocol/src/protocol.yml [hunk 8]
```diff
@@ -1191,6 +1191,22 @@ BrowserContext:
...
+        root: string
+        items:
+          type: array
+          items:
+            type: object
+            properties:
+              name: string
+              lastModifiedMs: number?
+      returns:
+        dir: string
```
Comment (dgozman) [inline packages/protocol/src/protocol.yml:1205]: Why does a client need to know a local path on the server?

File: packages/playwright-core/src/server/dom.ts (hunk 2 above)
Comment (mxschmitt) [reply packages/playwright-core/src/server/dom.ts:636]: I'm throwing on the client side instead, or would you prefer both for safety?

File: packages/playwright-core/src/server/dom.ts (hunk 4 above)
Comment (mxschmitt) [reply packages/playwright-core/src/server/dom.ts:650]: While file uploads to `input[webkitdirectory]` elements were possible before, this would mean we wait also for file uploads, while my change only waits for folder uploads. I prefer slightly only waiting for actual folder uploads, wdyt?

File: packages/protocol/src/protocol.yml (hunk 8 above)
Comment (mxschmitt) [reply packages/protocol/src/protocol.yml:1205]: it's a bit of a hack. We stream 3 files to the server (in the directory). The files itself could live under on the server:

- /tmp-random/my-dir/sub-dir/1.txt
- /tmp-random/my-dir/2.txt
- /tmp-random/my-dir/sub-dir/whatever/3.txt

So the server to this moment does not know what the root directory is. Instead we then send 3 files to the server over our protocol, get `/tmp-random/my-dir` back and pass it as a local-path to the server.

File: packages/playwright-core/src/client/elementHandle.ts (hunk 6 above)
Comment (mxschmitt) [reply packages/playwright-core/src/client/elementHandle.ts]: I simplified it a little bit but now there are a bunch of `if isDirectory` checks. I guess no way around a method. What about moving the map callback into an own method accepting the item, streams and localPath?

File: packages/protocol/src/protocol.yml (hunk 8 above)
Comment (dgozman) [reply packages/protocol/src/protocol.yml]: Yeah, we should not pass `localPaths` when talking to the server. Let's make server's temp files streams know what they need to.

File: packages/playwright-core/src/server/dom.ts (hunk 2 above)
Comment (dgozman) [reply packages/playwright-core/src/server/dom.ts:636]: Yeah, both sides sounds easier to understand.

File: packages/playwright-core/src/server/dom.ts (hunk 4 above)
Comment (dgozman) [reply packages/playwright-core/src/server/dom.ts:650]: I meant avoiding `fs.promises.stat` when `[webkitdirectory]` is not present. From what I see, we now always do this check unconditionally when passed a single file, which is the most common case.

Comment (mxschmitt) [reply packages/playwright-core/src/server/dom.ts:650]: If you want to throw https://github.com/microsoft/playwright/pull/31165#discussion_r1630015449 from both sides, this means we'd need to call stat from the client and server. So there is no way around that anyways?

File: packages/playwright-core/src/client/elementHandle.ts (hunk 6 above)
Comment (mxschmitt) [reply packages/playwright-core/src/client/elementHandle.ts]: Reverted to how it was before.

File: packages/playwright-core/src/server/dispatchers/browserContextDispatcher.ts [hunk 9]
```diff
@@ -178,13 +178,19 @@ export class BrowserContextDispatcher extends Dispatcher<BrowserContext, channel
...
     const tmpDir = path.join(dir, 'upload-' + createGuid());
-    await fs.promises.mkdir(tmpDir);
+    const tempDirWithRootName = params.rootDirName ? path.join(tmpDir, path.basename(params.rootDirName)) : tmpDir;
+    await fs.promises.mkdir(tempDirWithRootName, { recursive: true });
     this._context._tempDirs.push(tmpDir);
-    const file = fs.createWriteStream(path.join(tmpDir, params.name));
-    return { writableStream: new WritableStreamDispatcher(this, file, params.lastModifiedMs) };
+    return {
+      writableStreams: await Promise.all(params.items.map(async item => {
+        await fs.promises.mkdir(path.dirname(path.join(tempDirWithRootName, item.name)), { recursive: true });
```
[the remark this answers is not in the document]
Comment (mxschmitt) [reply packages/playwright-core/src/server/dispatchers/browserContextDispatcher.ts:190]: `item.name` can be in a sub-directory, hence we need it there.

File: packages/playwright-core/src/server/dom.ts [hunk 10]
```diff
@@ -625,29 +626,37 @@ export class ElementHandle<T extends Node = Node> extends js.JSHandle<T> {
...
       progress.throwIfAborted();  // Avoid action that has side-effects.
-      if (localPaths) {
-        await Promise.all(localPaths.map(localPath => (
+      if (localPaths || localDirectory) {
+        const localPathsOrDirectory = localDirectory ? [localDirectory] : localPaths!;
+        await Promise.all((localPathsOrDirectory).map(localPath => (
           fs.promises.access(localPath, fs.constants.F_OK)
         )));
-        await this._page._delegate.setInputFilePaths(retargeted, localPaths);
+        const waitForInputEvent = localDirectory ? this.evaluate(node => new Promise<any>(fulfill => {
```
Comment (dgozman) [inline packages/playwright-core/src/server/dom.ts:655]: Let's add a comment that we need this because browsers travers the directory asynchronously.

File: tests/electron/playwright.config.ts [hunk 11]
```diff
@@ -54,6 +54,7 @@ config.projects.push({
   name: 'electron-api',
   use: {
     browserName: 'chromium',
+    headless: false,
```
Comment (dgozman) [inline tests/electron/playwright.config.ts:57]: Seems unrelated.

File: packages/protocol/src/protocol.yml [hunk 12]
```diff
@@ -2187,6 +2196,7 @@ Frame:
               name: string
               mimeType: string?
               buffer: binary
+        localDirectory: string?
```
Comment (dgozman) [inline packages/protocol/src/protocol.yml:2199]: It seems like this should also include `directoryStream: WritableStream?` for remote scenarios, to make things obvious.

File: packages/playwright-core/src/client/elementHandle.ts [hunk 13]
```diff
@@ -262,18 +262,37 @@ export async function convertInputFiles(files: string | FilePayload | string[] |
   if (items.some(item => typeof item === 'string')) {
     if (!items.every(item => typeof item === 'string'))
       throw new Error('File paths cannot be mixed with buffers');
+    const itemFileTypes = (await Promise.all((items as string[]).map(async item => (await fs.promises.stat(item)).isDirectory() ? 'directory' : 'file')));
```
Comment (dgozman) [inline packages/playwright-core/src/client/elementHandle.ts:265]: I think it would be easier to split `items` into `files` and `directory` variables here, and not use `items` in the code below. WDYT?

File: packages/playwright-core/src/client/elementHandle.ts [hunk 14]
```diff
@@ -262,18 +262,37 @@ export async function convertInputFiles(files: string | FilePayload | string[] |
...
-      const streams: channels.WritableStreamChannel[] = await Promise.all((items as string[]).map(async item => {
-        const lastModifiedMs = (await fs.promises.stat(item)).mtimeMs;
-        const { writableStream: stream } = await context._wrapApiCall(() => context._channel.createTempFile({ name: path.basename(item), lastModifiedMs }), true);
-        const writable = WritableStream.from(stream);
-        await pipelineAsync(fs.createReadStream(item), writable.stream());
-        return stream;
-      }));
+      const streams = (await Promise.all((items).map(async item => {
+        const isDirectory = (await fs.promises.stat(item as string)).isDirectory();
+        const files = isDirectory ? (await fs.promises.readdir(item as string, { withFileTypes: true, recursive: true })).filter(f => f.isFile()).map(f => path.join(f.path, f.name)) : [item];
```
Comment (dgozman) [inline packages/playwright-core/src/client/elementHandle.ts:272]: This line and this whole method beg for more spacing and extra variables, or for an extracted method. It is pretty hard to follow now, not to say it wasn't easy even before 😄

File: tests/electron/playwright.config.ts (hunk 11 above)
Comment (mxschmitt) [reply tests/electron/playwright.config.ts:57]: Before Electron / WebView2 tests were treated as headless, so the `headless` fixture was `true` before and my check for the fixed upstream bug was not working as expected.

File: packages/playwright-core/src/server/fileUploadUtils.ts [hunk 15]
```diff
@@ -30,14 +30,17 @@ async function filesExceedUploadLimit(files: string[]) {
...
+  const { payloads, streams, directoryStream } = params;
+  let { localPaths, localDirectory } = params;
 
-  if ([payloads, localPaths, streams].filter(Boolean).length !== 1)
+  if ([payloads, localPaths, localDirectory, streams, directoryStream].filter(Boolean).length !== 1)
     throw new Error('Exactly one of payloads, localPaths and streams must be provided');
 
   if (streams)
-    localPaths = streams.map(c => (c as WritableStreamDispatcher).path());
+    localPaths = streams.map(c => (c as WritableStreamDispatcher)).map(s => s.path());
```
Comment (dgozman) [inline packages/playwright-core/src/server/fileUploadUtils.ts:40]:
```suggestion
    localPaths = streams.map(c => (c as WritableStreamDispatcher).path());
```

File: packages/playwright-core/src/server/fileUploadUtils.ts [hunk 16]
```diff
@@ -30,14 +30,17 @@ async function filesExceedUploadLimit(files: string[]) {
...
+  if (directoryStream)
+    localDirectory = (directoryStream as WritableStreamDispatcher | undefined)?.path();
```
Comment (dgozman) [inline packages/playwright-core/src/server/fileUploadUtils.ts:42]:
```suggestion
    localDirectory = (directoryStream as WritableStreamDispatcher).path();
```

**Review by dgozman — APPROVED** [review]
